### PR TITLE
Implement nested interpolations

### LINKED ISSUE #100: [Feature Request] resolve individual resolver arguments
It would be nice to enable recursive interpolation. 

Example : 

```python
import omegaconf
from omegaconf import OmegaConf

OmegaConf.register_resolver("plus", lambda x,y: float(x)+float(y))

c = OmegaConf.create({
        "i": 1,
        'key1': '${plus:1,2}',
        'key2': '${plus:${i},2}'
 })
print(c.key1)
# 3.0
print(c.key2)
# '${plus:1,2}'
```

But one would hope to get `3.0` out

Note that 2 recursion would be enough to enable the user to create deeper interpolation if they need it (by calling an other key which had already been interpolated).

### PR DESCRIPTION
**Pull request by odelalleau**
This is to fix #100 

Doc and tests are missing, I'm aware of it, I wanted to first get feedback on the implementation itself, in case there'd be any major concern.

The main parsing logic is in `_evaluate_complex()`, it's actually not too complex since we can sequentially iterate on all characters in the string and evaluate interpolations in the order we encounter their end `}`. So the core logic is simple but there are still "details" to be addressed (see concerns below).

This is not a super efficient implementation: (i) there are many string copies in the process, and (ii) the worst case complexity is quadratic in case of deep nesting (Edit: not anymore, see comment below). That being said, I don't expect it to be a real issue in the large majority of practical use cases.

A few potential concerns:
1. Although all tests currently pass, this can actually break backward compatibility since for instance if you had `a: 1` as config, then `${${a}}` used to evaluate to the string`"${1}"`, when now it would crash when trying to resolve this new interpolation. Hopefully people were not relying on this!
2. My current implementation returns a `None` value instead of crashing in case of syntax error. This isn't good, but right now the only two flags to trigger a crash are `throw_on_missing` and `throw_on_resolution_failure`, because there could be no syntax error in the previous code (for instance `${a}_${a` would be resolved into `"1_${a"`). Should I add a `throw_on_parsing_failure` flag?
3. With my current implementation `${a` resolves to `"${a"` while `${a}_${a` triggers a parsing failure. This inconsistency isn't good => I was thinking of using my new interpolation logic whenever it's not a simple interpolation but there is still a `${` in the string to be parsed, which should address this issue. Thoughts?
4. I'm not planning to allow escaping of the special `${` and `}` tokens, although it probably wouldn't be that hard. Let me know if you think this is important.

### CODE REVIEW

**Comment by odelalleau** [comment]
FYI here is a toy script I've been using for basic tests:

```
from omegaconf import OmegaConf

conf_str = """
a: 1
b: ${a} == 1
c: US
d: ER
e: ${env:${c}${d}}
f:
    US: 2
g: ${f.${c}} == 2
h: ${syntax_error.${a}
i: "{test}_${a}"
j: ${a}_${syntax_error
k: ${a
l: ${b} ${a
"""


def main():
    cfg = OmegaConf.create(conf_str)
    print(cfg.pretty())
    print(f"{cfg.b=}")
    print(f"{cfg.e=}")
    print(f"{cfg.g=}")
    print(f"{cfg.h=}")
    print(f"{cfg.i=}")
    print(f"{cfg.j=}")
    print(f"{cfg.k=}")
    print(f"{cfg.l=}")


if __name__ == "__main__":
    main()
```

and its output:

```
a: 1
b: ${a} == 1
c: US
d: ER
e: ${env:${c}${d}}
f:
  US: 2
g: ${f.${c}} == 2
h: ${syntax_error.${a}
i: '{test}_${a}'
j: ${a}_${syntax_error
k: ${a
l: ${b} ${a

cfg.b='1 == 1'
cfg.e='odelalleau'
cfg.g='2 == 2'
cfg.h=None
cfg.i='{test}_1'
cfg.j=None
cfg.k='${a'
cfg.l=None
```

**Comment by odelalleau** [comment]
Just had a flash that I could actually make the code both simpler & faster (now O(N)) => pushed 218bae71ca1e1e8d35696234ae6c2847cad10a01

File: omegaconf/_utils.py [hunk 1]
```diff
@@ -319,7 +319,7 @@ def get_value_kind(value: Any, return_match_list: bool = False) -> Any:
     MANDATORY_MISSING : "???
```
Comment (omry) [inline omegaconf/_utils.py:319]: fix quotes please

File: omegaconf/base.py [hunk 2]
```diff
@@ -4,10 +4,33 @@
...
 
+@dataclass
+class InterpolationRange:
+    """Used to store start/stop indices of interpolations being evaluated"""
+
+    start: int = -1
+    stop: int = -1
+
+
+class ParseError(Exception):
```
Comment (omry) [inline omegaconf/base.py:25]: Put it errors and inherit from OmegaConfBaseException

File: omegaconf/base.py [hunk 3]
```diff
@@ -4,10 +4,33 @@
...
+    """Thrown when unable to parse a complex interpolation"""
+
+    @staticmethod
+    def assert_(condition: Any) -> None:
+        if not condition:
+            raise ParseError()
```
Comment (omry) [inline omegaconf/base.py:31]: I don't think this logic belongs in the exception class.

File: omegaconf/base.py [hunk 4]
```diff
@@ -377,28 +400,63 @@ def _resolve_interpolation(
...
                 )
-                # if failed to resolve, return None for the whole thing.
-                if new_val is None:
-                    return None
-                new += orig[last_index : match.start(0)] + str(new_val)
-                last_index = match.end(0)
-
-            new += orig[last_index:]
+            except ParseError:
+                return None
```
Comment (omry) [inline omegaconf/base.py:413]: It it fly, if there is a need to suppress it in some scenarios we can add another flag.

File: omegaconf/base.py [hunk 5]
```diff
@@ -377,28 +400,63 @@ def _resolve_interpolation(
...
             return StringNode(value=new, key=key)
         else:
             assert False
 
+    def _evaluate_simple(self, value: str, **kw: Any) -> Optional["Node"]:
+        """Evaluate a simple interpolation"""
+        value_kind, match_list = get_value_kind(value=value, return_match_list=True)
```
Comment (omry) [inline omegaconf/base.py:420]: Can you reuse the previous parsing results here instead of re-parse?

File: omegaconf/base.py [hunk 6]
```diff
@@ -377,28 +400,63 @@ def _resolve_interpolation(
...
-                last_index = match.end(0)
-
-            new += orig[last_index:]
+            except ParseError:
+                return None
             return StringNode(value=new, key=key)
         else:
             assert False
 
+    def _evaluate_simple(self, value: str, **kw: Any) -> Optional["Node"]:
```
Comment (omry) [inline omegaconf/base.py:418]: why **kw and not explicit parameters?

File: omegaconf/base.py [hunk 7]
```diff
@@ -377,28 +400,63 @@ def _resolve_interpolation(
...
+        value_kind, match_list = get_value_kind(value=value, return_match_list=True)
+        assert value_kind == ValueKind.INTERPOLATION and len(match_list) == 1
+        match = match_list[0]
+        node = self._resolve_simple_interpolation(
+            inter_type=match.group(1), inter_key=match.group(2), **kw,
+        )
+        return node
+
+    def _evaluate_complex(self, value: str, **kw: Any) -> Optional[str]:
+        """Evaluate a complex interpolation (>1, nested, with other strings...)"""
```
Comment (omry) [inline omegaconf/base.py:429]: Please put a high level description of what this supports and how it works.
In the next major version I am planning to introduce a proper parser generator which will simplify this greatly.
It will help knowing what this function is supposed to be doing when integrating it.

**Comment by omry** [comment]
Please add proper tests, it's probably easier than doing it using your hack script.
see examples in the existing tests for pytest.mark.parameterize()

**Comment by omry** [comment]
be sure to include tests for custom interpolations containing nested interpolations as well.

File: omegaconf/base.py (hunk 5 above)
Comment (odelalleau) [reply omegaconf/base.py:420]: I could try but the problem is that the current parsing results are "incomplete", for instance if the string is `${f.${c}}` then it only matches `${c}`. So there could be two options:
1. Re-use the matches that have been found before: should be relatively easy to implement, but the code would be harder to read => not sure it's worth it
2. Improve the matching logic in `get_value_kind()` to fully cover all matches: this seems tricky, at least if we want to keep the current RE-based approach (https://stackoverflow.com/questions/133601/can-regular-expressions-be-used-to-match-nested-patterns)

File: omegaconf/base.py (hunk 6 above)
Comment (odelalleau) [reply omegaconf/base.py:418]: I use this pattern sometime when there are arguments that are not actually used by a function, but need to be provided so that they can be forwarded to another function. It's similar to the typical `__init__(self, arg1, arg2, **kw)` used with `super().__init__(**kw)` in some places. Main benefit is to avoid modifying the code in a lot of places whenever you need to modify some of these arguments. There are drawbacks though and I know some people don't like it, so I'll just change it here (will also be more consistent with the rest of the code).

File: omegaconf/_utils.py (hunk 1 above)
Comment (odelalleau) [reply omegaconf/_utils.py]: Fixed in fb08218

File: omegaconf/base.py (hunk 2 above)
Comment (odelalleau) [reply omegaconf/base.py]: Done in fb08218

File: omegaconf/base.py (hunk 3 above)
Comment (odelalleau) [reply omegaconf/base.py]: Changed in fb08218

File: omegaconf/base.py (hunk 4 above)
Comment (odelalleau) [reply omegaconf/base.py]: Changed in fb08218

I still hard to guard the exception within an `if throw_on_resolution_failure:` block as otherwise there would be an infinite recursion on error (because the code preparing the error message also tries to parse the same expression)

File: omegaconf/base.py (hunk 6 above)
Comment (odelalleau) [reply omegaconf/base.py]: Changed in 038f63c

File: omegaconf/base.py (hunk 7 above)
Comment (odelalleau) [reply omegaconf/base.py]: Done in ec534f9

**Comment by odelalleau** [comment]
NB: I added 540ba7d which solves an issue in my original implementation that was always returning a String node, even when the evaluated expression was another type.

File: omegaconf/base.py [hunk 8]
```diff
@@ -403,30 +398,65 @@ def _resolve_interpolation(
             try:
                 # NB: `match_list` is ignored here as complex interpolations may be
```
Comment (omry) [inline omegaconf/base.py:399]: what is NB ?

File: omegaconf/base.py (hunk 5 above)
Comment (omry) [reply omegaconf/base.py:422]: Okay, don't worry about it then.
Once we have a proper grammar this will be a million time easier.
in fact, check this out:
https://github.com/facebookresearch/hydra/blob/master/hydra/grammar/Override.g4#L78-L84

I think it will need to be changed to support complex interpolations from the command line.
In fact, I think it will need to be promoted from a token to a full fledged rule.

File: omegaconf/base.py [hunk 9]
```diff
@@ -403,30 +398,65 @@ def _resolve_interpolation(
...
-        """Evaluate a complex interpolation (>1, nested, with other strings...)"""
+    def _evaluate_complex(
+        self,
+        value: str,
+        key: Any,
+        throw_on_missing: bool,
+        throw_on_resolution_failure: bool,
+    ) -> Optional["Node"]:
+        """
+        Evaluate a complex interpolation (>1, nested, with other strings...).
```
Comment (omry) [inline omegaconf/base.py:442]: add a few examples for complex interpolations here.

File: omegaconf/base.py [hunk 10]
```diff
@@ -403,30 +398,65 @@ def _resolve_interpolation(
...
+
+        As an example, consider the string `value` set to "${foo.${bar}}":
+            1. We initialized our result with the original string: "${foo.${bar}}"
+            2. Scanning `value` from left to right, the first interpolation to be
+               closed is "${bar}". We evaluate it: if it resolves to "baz", we update
+               the result string to "${foo.baz}".
+            3. The next interpolation to be closed is now "${foo.baz}". We evaluate it:
+               if it resolves to "abc", we update the result accordingly to "abc".
+            4. We are done scanning `value` => the final result is "abc".
+        """
```
Comment (omry) [inline omegaconf/base.py:457]: Awesome description, thanks.

File: omegaconf/base.py (hunk 8 above)
Comment (odelalleau) [reply omegaconf/base.py:399]: Nota Bene (https://www.urbandictionary.com/define.php?term=NB)
Will replace with "Note"

File: omegaconf/base.py (hunk 5 above)
Comment (odelalleau) [reply omegaconf/base.py:422]: Yes a grammar will make things cleaner and more consistent. At least this PR should still enable some useful use cases in the meantime.

I ran into another issue related to the regex logic in `get_value_kind()`, which prevented me from fixing the inconsistency between `${a` (no error) and `${a}_${a ` (parsing error), which I had pointed out in point 3 above. It'd be a bit long to explain it in details, and annoying to fix, so I decided to give up on it for now since the grammar should fix it one way or another (= either always allowing it or always throwing an error).

**Comment by odelalleau** [comment]
> Please add proper tests, it's probably easier than doing it using your hack script.
> see examples in the existing tests for pytest.mark.parameterize()

Tests added in 5476596

File: omegaconf/base.py (hunk 8 above)
Comment (odelalleau) [reply omegaconf/base.py]: Done in 088f28a

**Comment by omry** [comment]
undraft when ready for final review.

File: omegaconf/base.py (hunk 9 above)
Comment (odelalleau) [reply omegaconf/base.py]: Done in 46f6d60

File: omegaconf/errors.py [hunk 11]
```diff
@@ -103,3 +103,9 @@ class ConfigValueError(OmegaConfBaseException, ValueError):
     """
     Thrown from a config object when a regular access would have caused a ValueError.
     """
+
+
+class ParseError(OmegaConfBaseException):
+    """
+    Thrown when unable to parse a complex interpolation.
+    """
```
[the remark this answers is not in the document]
Comment (odelalleau) [reply omegaconf/errors.py]: Done in 67231e0

**Comment by odelalleau** [comment]
> undraft when ready for final review.

Should be ready :)

A note about additional commits I haven't mentioned in the above comments:
* 5fa85ef fixed an issue caused by the fact I was assuming all `${..}` segment were interpolations, while currently this is not the case (I guess that might change later with the grammar, but I preferred not to touch the regex in `get_value_kind()` for now)
* 2f99a64 adds documentation

**Comment by omry** [comment]
A tiny coverage issue, but still 99.99% is not 100% :)

File: docs/notebook/Tutorial.ipynb [hunk 12]
```diff
@@ -731,7 +731,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "You can add additional interpolation types using custom resolvers. This example creates a resolver that adds 10 the the given value."
+    "You can add additional interpolation types using custom resolvers, which take strings as inputs. This example creates a resolver that adds 10 to the given value (note the need to cast it to `int`)."
```
Comment (omry) [inline docs/notebook/Tutorial.ipynb:734]: I will improve this when I have a proper grammar. for now this is indeed taking strings which is unfortunate.

File: docs/notebook/Tutorial.ipynb [hunk 13]
```diff
@@ -757,6 +757,40 @@
...
+      "3\n"
+     ]
+    }
+   ],
+   "source": [
+    "OmegaConf.register_resolver(\"plus_int\", lambda x, y: int(x) + int(y))\n",
+    "conf = OmegaConf.create({\"a\": 1, \"b\": 2, \"a_plus_b\": \"${plus_int:${a},${b}}\"})\n",
+    "print(conf.a_plus_b)"
+   ]
+  },
```
Comment (omry) [inline docs/notebook/Tutorial.ipynb:793]: Can you also add an example for a simple nested interpolation?
something along the lines of:
```python
cfg = OmegaConf.create({
  "plans":{
    "A": "plan A",
    "B": "plan B",
  },
  "selected_plan": "A"
  "plan": "${plans.${selected_plan}}"
})

cfg.plan
>> "Plan A"
cfg.selected_plan = "B"
cfg.plan
>> "Plan B"
```

File: docs/source/usage.rst [hunk 14]
```diff
@@ -351,6 +351,18 @@ You can use literal commas and spaces anywhere by escaping (:code:`\,` and :code
     >>> c.escape_whitespace
     'Hello World'
 
+Nested interpolations
```
Comment (omry) [inline docs/source/usage.rst:354]: added the example for simple nested interpolation here as well.

File: omegaconf/base.py (hunk 5 above)
Comment (omry) [reply omegaconf/base.py:422]: Please explain the problem more and how a user would hit this.
I like this change but I want to make sure it does not opening up significant issues we are not ready to deal with.
It will be MANY months before OmegaConf 2.1.

File: omegaconf/base.py [hunk 15]
```diff
@@ -439,15 +449,22 @@ def _evaluate_complex(
         throw_on_resolution_failure: bool,
     ) -> Optional["Node"]:
         """
-        Evaluate a complex interpolation (>1, nested, with other strings...).
+        Evaluate a complex interpolation.
+
+        A complex interpolation is more elaborate than "${a}" or "${a:b,c}", e.g.:
+            "I really like ${liked}" (concatenating a string)
+            "${subject} {verb} ${object}" (concatenating multiple strings)
```
Comment (omry) [inline omegaconf/base.py:456]: those two are referred to everywhere else as string interpolations. This is actually user facing terminology:

see [this](https://docs.google.com/presentation/d/e/2PACX-1vT_UIV7hCnquIbLUm4NnkUpXvPEh33IKiUEvPRF850WKA8opOlZOszjKdZ3tPmf8u7hGNP6HpqS-NT5/pub?start=false&loop=false&delayms=3000&slide=id.g84632f636b_5_148).

You are adding support for a new dimension, which is nested interpolations.
this is orthogonal to simple (`${foo.bar}` or string (`${foo}_bar`)) interpolations. both can be nested or not.

File: omegaconf/base.py [hunk 16]
```diff
@@ -523,10 +540,10 @@ def _re_parent(self) -> None:
 
 def _parse_assert(condition: Any, msg: str = "") -> None:
```
Comment (omry) [inline omegaconf/base.py:541]: this method does not parse anything, certainly not an assert ;)

maybe `_cond_parse_error(condition: Any, msg: str = "")`?

File: tests/test_interpolation.py [hunk 17]
```diff
@@ -347,11 +348,99 @@ def test_incremental_dict_with_interpolation() -> None:
...
+        for key, expected in expected_dict.items():
+            assert OmegaConf.select(c, key) == expected
+    finally:
+        OmegaConf.clear_resolvers()
+
+
+@pytest.mark.parametrize(  # type: ignore
+    "cfg,key",
+    [
+        ({"a": "PATH", "b": "${env:${a}"}, "b"),
```
Comment (omry) [inline tests/test_interpolation.py:433]: Why is this an error?

File: omegaconf/base.py (hunk 5 above)
Comment (odelalleau) [reply omegaconf/base.py:422]: With both the current master and this PR, you can use:
```
a: 1
b: ${a
```
and it will run just fine, with `conf.b` being equal to `"${a"`.

However, with this PR, if you do:
```
a: 1
b: x_${a
```
then trying to access `conf.b` will trigger a parsing error. This is because by adding `x_` as prefix, it's not a simple interpolation anymore, so it goes through my new parsing code that ensures that all opening braces are properly closed.

Ideally I would like the first case to also trigger an error. I don't think anyone should be relying on the fact that right now it "works". But I don't see an easy way to make it happen.

So this is not really adding a problem, it's more that I'm unable to fix an existing problem :)

File: docs/notebook/Tutorial.ipynb (hunk 12 above)
Comment (odelalleau) [reply docs/notebook/Tutorial.ipynb:734]: Cool, I just added this comment because it's easy to do something like `${plus:1,2}` and end up with `"12"` instead of `3` :)

File: docs/notebook/Tutorial.ipynb (hunk 13 above)
Comment (odelalleau) [reply docs/notebook/Tutorial.ipynb:793]: Good idea! Done in a15623d

File: docs/source/usage.rst (hunk 14 above)
Comment (odelalleau) [reply docs/source/usage.rst:354]: Also in a15623d

File: omegaconf/base.py (hunk 15 above)
Comment (odelalleau) [reply omegaconf/base.py]: Updated docstring in 3170517

Note that `_evaluate_complex()` is now also used for string interpolations (that's why I used "complex" instead of "nested").

File: omegaconf/base.py (hunk 5 above)
Comment (omry) [reply omegaconf/base.py:422]: This is fine. It's good that you are noticing these differences but don't try too hard to preserve 100% compatibility.
some small differences around error cases are fine.

File: omegaconf/base.py (hunk 15 above)
Comment (odelalleau) [reply omegaconf/base.py]: Hmm, your comment about simple interpolation being potentially nested makes me think I need to to a renaming pass over the code though, because `_resolve_simple_interpolation()` doesn't support nesting. I had interpreted "simple" as "a single `${..}` block" until now. This will have to wait until next week.

File: tests/test_interpolation.py (hunk 17 above)
Comment (odelalleau) [reply tests/test_interpolation.py:433]: Oops, renamed in 94474ae
I intended it to be understood as "parsing_assert" but yeah, that was a bad name :)

File: omegaconf/base.py (hunk 16 above)
Comment (odelalleau) [reply omegaconf/base.py]: Oops, renamed in 94474ae
I intended it to be understood as "parsing_assert" but yeah, that was a bad name :)

File: tests/test_interpolation.py (hunk 17 above)
Comment (odelalleau) [reply tests/test_interpolation.py:433]: Missing closing brace (I'm glad you missed it, it means such an error is useful :) )

Comment (omry) [reply tests/test_interpolation.py:434]: sneaky sneaky error.

**Comment by omry** [comment]
almost there. You have some ci issues. You can run it locally with nox.

File: omegaconf/base.py (hunk 15 above)
Comment (omry) [reply omegaconf/base.py]: Yeah, maybe take a break and come back for a consistency pass later.

**Comment by odelalleau** [comment]
> almost there. You have some ci issues. You can run it locally with nox.

Should be fixed now (2b473fe) -- taking note to run nox locally next time :)

**Comment by omry** [comment]
Did a full review and added a few more comments.

This is looking great, the only thing that makes me pause here is that generally I do not add new features for minor versions.
This should be in OmegaConf 2.1, so this entire thing would probably go into a new branch for 2.1.
we can have a 2.1.0 pre release package to make it easier to install.

This reminds me, please add a news fragment (see [this](https://hydra.cc/docs/next/development/contributing/#news-entries) this for context).

File: docs/notebook/Tutorial.ipynb [hunk 18]
```diff
@@ -757,6 +757,74 @@
     "print(conf.key)"
    ]
   },
+  {
```
Comment (omry) [inline docs/notebook/Tutorial.ipynb:760]: Looking at this diff is making it hard to know the structure is right.
The structure that makes sense to me:

1. Variable interpolation section contains a mention and an example of nesting interpolations.
2. Custom interpolation mentions that nested interpolations are supported, providing no example or a very short one.

File: docs/source/usage.rst [hunk 19]
```diff
@@ -351,6 +351,35 @@ You can use literal commas and spaces anywhere by escaping (:code:`\,` and :code
...
+^^^^^^^^^^^^^^^^^^^^^
```
Comment (omry) [inline docs/source/usage.rst:355]: same feedback as the one for the notebook:
Nested interpolations applies to all interpolations, they should be documented before custom interpolations.
I am a bit torn about mentioning nested interpolations in custom interpolations once we have a general description.

Also use the term "nested interpolations" or "interpolations can be nested".

File: docs/notebook/Tutorial.ipynb (hunk 18 above)
Comment (odelalleau) [reply docs/notebook/Tutorial.ipynb:760]: Question related to this, currently the section structure in the notebook is like this:
```
# Variable interpolation
### Config node interpolation
### Environment variable interpolation
## Custom interpolations
## Nested interpolations
```
I assume that the first two sub-sections should use `##` instead, am I correct?

**Comment by odelalleau** [comment]
> This reminds me, please add a news fragment (see [this](https://hydra.cc/docs/next/development/contributing/#news-entries) this for context).

Added news in edbb09b

File: omegaconf/_utils.py [hunk 20]
```diff
@@ -316,10 +316,10 @@ def get_value_kind(value: Any, return_match_list: bool = False) -> Any:
     """
     Determine the kind of a value
     Examples:
-    MANDATORY_MISSING : "???
+    MANDATORY_MISSING : "???"
     VALUE : "10", "20", True,
     INTERPOLATION: "${foo}", "${foo.bar}"
```
[the remark this answers is not in the document]
Comment (odelalleau) [reply omegaconf/_utils.py:321]: I'm afraid not: this is related to the naming issue discussed above (the meaning of "simple" interpolation), which will take a bit of time to fix properly. Currently `get_value_kind()` returns `STR_INTERPOLATION` for a nested simple interpolation, and this is required by my implementation.

File: omegaconf/_utils.py [hunk 21]
```diff
@@ -709,6 +709,17 @@ def type_str(t: Any) -> str:
         return ret
 
 
+def update_string(to_s: str, start: int, stop: int, from_s: str) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (odelalleau) [reply omegaconf/_utils.py]: Changed in a97ddfd

File: tests/test_interpolation.py [hunk 22]
```diff
@@ -347,11 +348,100 @@ def test_incremental_dict_with_interpolation() -> None:
...
+    "cfg,expected_dict",
+    [
+        (
+            """
+            a: 1
+            b: a
+            c: ${${b}}
+            """,
+            {"c": 1},  # basic nesting
+        ),
```
[the remark this answers is not in the document]
Comment (odelalleau) [reply tests/test_interpolation.py]: Done in 7ed9492

File: tests/test_utils.py [hunk 23]
```diff
@@ -485,3 +485,21 @@ def test_is_list_annotation(type_: Any, expected: Any) -> Any:
...
+    [
+        (-1, 1),
+        (1, -1),
+        (2, 1),
+        (len("hello world") + 1, 100),
+        (0, len("hello world") + 1),
+    ],
+)
+def test_update_string_error(start: int, stop: int) -> None:
+    assert _utils.update_string("hello world", start, stop, "i") is None
```
[the remark this answers is not in the document]
Comment (odelalleau) [reply tests/test_utils.py]: Done in a97ddfd

File: docs/notebook/Tutorial.ipynb (hunk 18 above)
Comment (odelalleau) [reply docs/notebook/Tutorial.ipynb:794]: Changed in 6339fcd

File: docs/source/usage.rst (hunk 19 above)
Comment (odelalleau) [reply docs/source/usage.rst]: Changed in 6339fcd

For now I kept the example in custom interpolations as personally I think it's a typical use case that's worth highlighting.

**Comment by odelalleau** [comment]
Just FYI nox session `coverage-3.6` passed on Github but failed on my Macbook. Seems to be related to the `id=` parameter of `pytest.param()`. Not sure if I should just revert that change (?)

File: docs/notebook/Tutorial.ipynb (hunk 18 above)
Comment (omry) [reply docs/notebook/Tutorial.ipynb:794]: Probably, you can see what it looks like in the local browser by building the docs and opening the index locally.
Please do that do to make sure it all looks good (there are some details on how to do it at the bottom of [CONTRIBUTING.md](https://github.com/omry/omegaconf/blob/master/CONTRIBUTING.md).

File: docs/source/usage.rst [hunk 24]
```diff
@@ -286,6 +286,23 @@ Example:
...
+    ...     A: plan A
+    ...     B: plan B
+    ... selected_plan: A
+    ... plan: ${plans.${selected_plan}}
+    ... """)
+    >>> print(f"Default: cfg.plan = {cfg.plan}")
+    Default: cfg.plan = plan A
+    >>> cfg.selected_plan = "B"
+    >>> print(f"After selecting plan B: cfg.plan = {cfg.plan}")
+    After selecting plan B: cfg.plan = plan B
```
Comment (omry) [inline docs/source/usage.rst:304]:
```suggestion
    >>> print(cfg.plan) # default plan
    plan A
    >>> cfg.selected_plan = "B"
    >>> print(cfg.plan) # new plan
    plan B
```

File: omegaconf/_utils.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (omry) [reply omegaconf/_utils.py:321]: Please add a comment calling it out. This is a subtle but important details for this function.

File: omegaconf/base.py [hunk 25]
```diff
@@ -5,7 +5,20 @@
...
+    UnsupportedInterpolationType,
+)
+
+
+@dataclass
+class InterpolationRange:
+    """Used to store start/stop indices of interpolations being evaluated"""
+
+    start: int = -1
+    stop: int = -1
```
Comment (omry) [inline omegaconf/base.py:21]: could it make more sense to use Optional[int] instead of -1 here?

File: tests/test_interpolation.py [hunk 26]
```diff
@@ -347,11 +348,106 @@ def test_incremental_dict_with_interpolation() -> None:
...
+    "cfg,expected_dict",
+    [
+        pytest.param(
+            """
+            a: 1
+            b: a
+            c: ${${b}}
+            """,
+            {"c": 1},
+            id="basic nesting",
```
Comment (omry) [inline tests/test_interpolation.py:366]: Please avoid special characters in ids, no spaces, no ().
(While this works, it confuses other tools like Pycharm).

File: tests/test_interpolation.py [hunk 27]
```diff
@@ -347,11 +348,106 @@ def test_incremental_dict_with_interpolation() -> None:
...
+            assert OmegaConf.select(c, key) == expected
+    finally:
+        OmegaConf.clear_resolvers()
+
+
+@pytest.mark.parametrize(  # type: ignore
+    "cfg,key",
+    [
+        # All these examples have non-matching braces.
+        ({"a": "PATH", "b": "${env:${a}"}, "b"),
```
Comment (omry) [inline tests/test_interpolation.py:440]: can also add pytest.param with ids here.

**Comment by odelalleau** [comment]
> Since it passes in CI it's okay.
> I did not see this before, can you show me the failure?

So I was wrong, this has nothing to do with adding `id=`, these errors actually occur in the `master` branch for me. Here's a prototypical example:

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_base_config.py:480: in <lambda>
    @pytest.mark.parametrize("copy_method", [lambda x: copy.copy(x), lambda x: x.copy()])
omegaconf/listconfig.py:310: in copy
    return copy.copy(self)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = [1, 2, '${0}']

    def copy(x):
        """Shallow copy operation on arbitrary Python objects.

        See the module's __doc__ string for more info.
        """

        cls = type(x)

        copier = _copy_dispatch.get(cls)
        if copier:
            return copier(x)

        try:
            issc = issubclass(cls, type)
        except TypeError: # cls is not a class
            issc = False
        if issc:
            # treat it as a regular class:
            return _copy_immutable(x)

        copier = getattr(cls, "__copy__", None)
        if copier:
>           return copier(x)
E           TypeError: __copy__() takes 1 positional argument but 2 were given
```

I haven't really tried to figure out the root cause. Just did a quick test with a conda Python 3.6 env and I can't repro. But it systematically fails with nox...

File: omegaconf/_utils.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (odelalleau) [reply omegaconf/_utils.py:321]: Comment added in e818f78

File: docs/notebook/Tutorial.ipynb (hunk 18 above)
Comment (odelalleau) [reply docs/notebook/Tutorial.ipynb:794]: Checked, structure looks good to me.

I realized though that one example was too wide, at least on my computer (had to scroll to the right to see the full line of code). I fixed that in 5500100

File: docs/source/usage.rst (hunk 24 above)
Comment (odelalleau) [reply docs/source/usage.rst]: Took your change (d50df1a)

File: omegaconf/base.py (hunk 25 above)
Comment (odelalleau) [reply omegaconf/base.py]: Done in b3bedcb (actually making `start` a mandatory `int` to simplify type checks)

File: tests/test_interpolation.py (hunk 26 above)
Comment (odelalleau) [reply tests/test_interpolation.py]: Done in f25f17f

File: tests/test_interpolation.py (hunk 27 above)
Comment (odelalleau) [reply tests/test_interpolation.py]: Done in f25f17f

File: omegaconf/base.py (hunk 25 above)
Comment (odelalleau) [reply omegaconf/base.py]: Done in b3bedcb (actually making start a mandatory int to simplify type checks)

**Comment by omry** [comment]
> > Since it passes in CI it's okay.
> > I did not see this before, can you show me the failure?
> 
> So I was wrong, this has nothing to do with adding `id=`, these errors actually occur in the `master` branch for me. Here's a prototypical example:
> ...
> 
> I haven't really tried to figure out the root cause. Just did a quick test with a conda Python 3.6 env and I can't repro. But it systematically fails with nox...

Can you file an issue with a minimal repro? it's a bit hard to parse here.

**Comment by odelalleau** [comment]
> Can you file an issue with a minimal repro? it's a bit hard to parse here.

See #314 and associated PR #315

**Comment by odelalleau** [comment]
Just a note that I rebased on top of master to be sure everything still worked fine

**Comment by omry** [comment]
Awesome work, I am ready to land this.
See my final comment (that was stuck in pending for 2 days :) ).

**Comment by omry** [comment]
lending this to be able to test things on your followup issue. Please respond to my comment.

File: omegaconf/_utils.py [hunk 28]
```diff
@@ -316,10 +316,15 @@ def get_value_kind(value: Any, return_match_list: bool = False) -> Any:
...
+    MANDATORY_MISSING : "???"
     VALUE : "10", "20", True,
     INTERPOLATION: "${foo}", "${foo.bar}"
-    STR_INTERPOLATION: "ftp://${host}/path"
+    STR_INTERPOLATION: "ftp://${host}/path", "${foo.${bar}}"
+
+    Note in particular that in the current implementation, "${foo.${bar}}" is
+    identified as a string interpolation (`STR_INTERPOLATION`) while it is
+    actually a (nested) simple interpolation. This discrepancy w.r.t naming
+    conventions will be addressed in a future update.
```
[the remark this answers is not in the document]
Comment (odelalleau) [reply omegaconf/_utils.py:327]: Yes, please assign #319 to me
Thanks for landing :)
